internal/controller/supervisor.go

```diff
@@ -0,0 +1,140 @@
+package controller
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/consul/lib/retry"
+)
+
+// flapThreshold is the minimum amount of time between restarts for us *not* to
+// consider a controller to be stuck in a crash-loop.
+const flapThreshold = 2 * time.Second
```

Comment:
This value was plucked out of thin air 💨 maybe something longer would be more appropriate, but then we'd need to make it configurable for tests.